### PR TITLE
LTI Access Token module

### PR DESCRIPTION
**Pull request by carl-codeorg**
Adds a module for retrieving a JWT from an LTI platform. This is the JWT used for calling LTI Advantage APIs on behalf of the code.org LTI integration. This adds the module, but it isn't yet imported anywhere in the code. The first feature to take advantage of the module will be LMS roster syncing.

## Links

[JIRA](https://codedotorg.atlassian.net/browse/P20-314)

## Testing story

Added some unit tests - more testing can be put in place after this module is incorporated into the [rostering functionality](https://codedotorg.atlassian.net/browse/P20-405).

This can be tested locally against the Canvas sandbox as long as you have the LTI integration configured both in the sandbox and locally. Ping me for details on setting this up.

## Caching

This uses the CDO shared_cache to store JWTs. Cached tokens can be reused for any call to LTI APIs for a given integration.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by carl-codeorg** [comment]
Some tests failing that weren't failing locally. Will take a look

File: dashboard/lib/lti_access_token.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
+require 'json'
+module LtiAccessToken
+  # Retrieve a JWT from the access_token_url of the LTI Integration
+  # The JWT can be used to access the LTI Advantage Services for the LTI Integration
+  # client_id and issuer are present in the id_token we cache or store as a cookie from the LTI launch
+  def get_access_token(client_id, issuer, scopes = Policies::Lti::ALL_SCOPES)
+    cache_key = "#{Policies::Lti::NAMESPACE}/#{client_id}-#{issuer}"
+
+    # check for valid access token
+    access_token = CDO.shared_cache.read(cache_key)
```
Comment (daynew) [inline dashboard/lib/lti_access_token.rb:10]: I know the spec details using the cache for this, but do you know how long the expirations usually are? I'm wondering how much data we might end up caching and how that compares to how much memory our current memcached server has.

File: dashboard/lib/lti_access_token.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    # cache access_token, set expiration
+    CDO.shared_cache.write(cache_key, access_token, expires_in: exp)
+
+    return access_token
+  end
+
+  # Takes a hash of claims and returns a signed JWT.
+  # See https://datatracker.ietf.org/doc/html/rfc7519#section-4.1
+  # for definitions of the claims.
+  def sign_jwt(payload, private_key_json = CDO.jwk_private_key_data)
```
Comment (daynew) [inline dashboard/lib/lti_access_token.rb:52]: Should we throw an exception if `CDO.jwk_private_key_data` is not defined so the developer knows there is a configuration issue?

File: dashboard/test/lib/lti_access_token_test.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    @fake_access_token_response = OpenStruct.new({body: JSON.generate(fake_response_hash)})
+  end
+
+  test 'get_access_token calls the access_token_url from the LTI integration' do
+    HTTParty.stubs(:post).with(@lti_integration.access_token_url, anything, anything).returns(@fake_access_token_response)
+    jwt = get_access_token(@lti_integration.client_id, @lti_integration.issuer, ['test_scope', 'test_scope2'])
+    assert_equal 'fake_access_token', jwt
+  end
+
+  test 'get_access_token uses cached tokens if present' do
```
Comment (daynew) [inline dashboard/test/lib/lti_access_token_test.rb:38]: We should probably also have a unit tests which exercises the expiration too. Look at other unit tests which freeze and move the clock forward a specific amount of time. I think I did some for CPA.

**Review by daynew — APPROVED** [review]
Looking good! I spent a lot of time reading LTI stuff and understanding how this fits into the whole puzzle.

File: dashboard/lib/lti_access_token.rb (hunk 1 above)
Comment (carl-codeorg) [reply dashboard/lib/lti_access_token.rb:10]: Yeah it's a good point. Canvas defaults to 1-hour expirations, which I think is common, but we might see shorter or longer expirations from other LMS. For now we should be good, considering it'll be one token cached per client_id/issuer pair, but I'm also checking with infra re: memcached costs.

**Pull request merged**